### DISCUSSION TITLE
tokio::time::sleep blocks other requests

### DISCUSSION THREAD

**Discussion by valery-kirichenko** [tokio-rs/axum#2436] [Q&A] [answered]
### Summary

I am trying to simulate requests taking some time to complete using `tokio::time::sleep` in the request handler but it looks like it blocks the whole thread instead of allowing other requests to be processed concurrently.
 
Here's my code:
```rust
use std::time::Duration;
use axum::Router;
use axum::routing::get;

#[tokio::main]
async fn main() {
    let app = Router::new()
        .route("/", get(root));

    let listener = tokio::net::TcpListener::bind("127.0.0.1:8082").await.unwrap();
    axum::serve(listener, app).await.unwrap();
}

async fn root() -> &'static str {
    tokio::time::sleep(Duration::from_secs(10)).await;
    "Hello, World!"
}
```

Run it, [open](http://127.0.0.1:8082) 2 tabs in the browser. I expect both of them to load in ~10 seconds. Instead, it takes 10 seconds to load the first one and almost 20 seconds to load the second one (depending on how quickly you open second tab after the first one).

Why does it behave like that? I tried searching and all I've seen is that it should occur when you (incorrectly) use `std::thread::sleep` with recommendations to use `tokio::time::sleep`.

axum version 0.7.2, tokio version 1.35.0

**Comment by jplatte** [comment, marked as answer]
Your browser won't create two independent connections to the same host, it serializes the requests. This is not the first time people have stumbled upon this and created an axum issue / discussion, but if you try with browser + curl, or two curl commands you should see that the server is behaving normally.
